torch_xla/csrc/runtime/cache.h

```diff
@@ -106,6 +122,116 @@ class Cache {
+        readonly_(readonly),
+        serialize_(serialize),
+        deserialize_(deserialize) {
+    std::filesystem::create_directories(cache_dir);
+  }
+
+  // Add the value to the persistent cache. This only writes to disk if no
+  // existing value is tracked to avoid unnecessary serialization overhead.
+  // The value will also be tracked in memory, so subsequent Get calls will not
+  // incur deserialization.
+  // If the cache is readonly, nothing is written to disk.
+  TypePtr Add(K key, TypePtr obj) override {
+    std::lock_guard<std::mutex> slock(lock_);
+    std::string path = GetPath(key);
+    if (!Exists(path) && !readonly_) {
+      std::ofstream out(path, std::ios::binary);
+      out << serialize_(obj);
+    }
+    return memory_cache_.Add(key, obj);
+  }
+
+  // Get the TypePtr associated with the key. This method will first check
+  // if the key is tracked in memory, and if not it will check for a persisted
+  // version on disk.
+  TypePtr Get(const K& key) override {
+    std::lock_guard<std::mutex> slock(lock_);
+    TypePtr mem = memory_cache_.Get(key);
+    if (mem) {
+      return mem;
+    }
+
+    std::string path = GetPath(key);
+    if (!Exists(path)) {
+      TORCH_LAZY_COUNTER("PersistentCacheMiss", 1);
+      return nullptr;
+    }
+    std::stringstream ss;
+    std::ifstream in(path, std::ios::binary);
+    ss << in.rdbuf();
+    std::string serialization = ss.str();
+
+    TypePtr val = deserialize_(serialization);
+    if (!val) {
+      TORCH_LAZY_COUNTER("PersistentCacheDeserializeFailure", 1);
+      // Remove the serialized value from disk to allow a new value to be stored
+      _erase(key);
+      return nullptr;
+    }
+    TORCH_LAZY_COUNTER("PersistentCacheHit", 1);
+    // Make sure the memory_cache_ tracks the value to prevent multiple loads
+    return memory_cache_.Add(key, val);
+  }
+
+  void Clear() override {
+    std::lock_guard<std::mutex> slock(lock_);
+    memory_cache_.Clear();
+    // Delete and recreate the cache directory on disk.
+    if (!readonly_) {
+      std::filesystem::remove_all(cache_dir_);
+      std::filesystem::create_directories(cache_dir_);
+    }
+  }
+
+  bool Erase(const K& key) override {
+    std::lock_guard<std::mutex> slock(lock_);
+    return _erase(key);
+  }
+
+  Cache<K, T, H, E>& GetMemoryCache() { return memory_cache_; }
+
+ private:
+  std::string GetPath(K key) {
+    std::stringstream ss;
+    ss << key;
+    return cache_dir_ / std::filesystem::path(ss.str());
+  }
+
+  bool Exists(std::string path) {
+    struct stat buffer;
+    return stat(path.c_str(), &buffer) == 0;
+  }
+
+  bool _erase(const K& key) {
+    memory_cache_.Erase(key);
+    return !readonly_ && std::filesystem::remove(GetPath(key));
+  }
+
+  Cache<K, T, H, E> memory_cache_;
+  std::function<std::string(TypePtr&)> serialize_;
+  std::function<TypePtr(std::string&)> deserialize_;
+  std::filesystem::path cache_dir_;
+  std::mutex lock_;
+  bool readonly_;
```

Comment:
Can we make `readonly_` to be constant?

Reply:
It seems readonly_ means the persistent cache is readonly but the memory cache is rw... And it confuses me as I thought the whole cache is readonly. Can we have more documentation about the flag?

Reply:
I see, that's a good caveat - readonly refers only to the persistence layer. I'll add some documentation around this.